File: backend/rpc/handler/user/email.py
from flask import request

from rpc.gen.user.email.services import TUserEmailService
from rpc.gen.user.email.errors.ttypes import TEmailError, TEmailErrorCode

from models import get_model
from converter.user import DBUser_TUser

class TUserEmailServiceHandler(TUserEmailService.Iface):
    def __init__(self):
        self.auth_model = get_model('auth')
        self.user_model = get_model('user')
        self.email_model = get_model('email')

    def has_password(self, email_token):
        ip = request.remote_addr
        email_payload = self.email_model.decode_verify(ip, email_token)
        db_user = self.user_model.get_user(email_payload['user_id'])
        if db_user.email_secret_2 != email_payload['email_secret_2']:
            raise TEmailError(TEmailErrorCode.EMAIL_TOKEN_EXPIRED)
        return db_user.has_password

    def verify_email(self, verify_token, password):
        ip = request.remote_addr
        email_payload = self.email_model.decode_verify(ip, verify_token)
        db_user = self.user_model.verify_email(
            email_payload['user_id'],
            email_payload['email_secret_2'],
            email_payload['email'],
            password
        )

    def change_email(self, verify_token, password):
        ip = request.remote_addr
        email_payload = self.email_model.decode_verify(ip, verify_token)
        db_user = self.user_model.change_email(
            email_payload['user_id'],
            email_payload['email_secret_2'],
            email_payload['email'],
            password
        )

    def set_password(self, password_token, password):
        ip = request.remote_addr
        password_payload = self.email_model.decode_password(ip, password_token)
        db_user = self.user_model.reset_password(
            password_payload['user_id'],
            password_payload['email_secret_2'],
            password
        )
        

File: frontend/src/rpc/client/user/email.js
import TUserEmailService from '@/rpc/gen/TUserEmailService';
import { TBaseClient } from '@/rpc/client/base';


class TUserEmailClient extends TBaseClient{
	constructor(stores=null){
		super(stores, TUserEmailService, '/api/user/email');
	}
	async has_password(token){
		//this.stores.helper.auth.requireLogout();
		return await this.client.has_password(token);
	}

	async verify_email(token, password){
		//this.stores.helper.auth.requireLogout();
		await this.client.verify_email(token, password);
		if (this.stores.auth.isLoggedIn){
			await this.stores.client.user.profile.get_user();
		}
	}

	async change_email(token, password){
		//this.stores.helper.auth.requireLogout();
		await this.client.change_email(token, password);
		if (this.stores.auth.isLoggedIn){
			await this.stores.client.user.profile.get_user();
		}
	}
	async set_password(token, password){
		//this.stores.helper.auth.requireLogout();
		await this.client.set_password(token, password);
		if (this.stores.auth.isLoggedIn){
			await this.stores.client.user.auth.login(
				this.stores.auth.user.username,
				password
			);
		}
	}
}

const userEmailClient = new TUserEmailClient();

export { TUserEmailClient, userEmailClient }
export default userEmailClient